libp2p/switch.nim

```diff
@@ -325,24 +325,31 @@ proc subscribeToPeer(s: Switch, peerInfo: PeerInfo) {.async, gcsafe.} =
       warn "unable to initiate pubsub", exc = exc.msg
       s.dialedPubSubPeers.excl(peerInfo.id)
 
-proc subscribe*(s: Switch, topic: string, handler: TopicHandler): Future[void] {.gcsafe.} =
+proc subscribe*(s: Switch, topic: string,
+                handler: TopicHandler): Future[void] =
   ## subscribe to a pubsub topic
   if s.pubSub.isNone:
-    raise newNoPubSubException()
+    var retFuture = newFuture[void]("Switch.subscribe")
+    retFuture.fail(newNoPubSubException())
+    return retFuture
 
   result = s.pubSub.get().subscribe(topic, handler)
 
-proc unsubscribe*(s: Switch, topics: seq[TopicPair]): Future[void] {.gcsafe.} =
+proc unsubscribe*(s: Switch, topics: seq[TopicPair]): Future[void] =
   ## unsubscribe from topics
   if s.pubSub.isNone:
-    raise newNoPubSubException()
+    var retFuture = newFuture[void]("Switch.unsubscribe")
+    retFuture.fail(newNoPubSubException())
+    return retFuture
 
   result = s.pubSub.get().unsubscribe(topics)
 
-proc publish*(s: Switch, topic: string, data: seq[byte]): Future[void] {.gcsafe.} =
+proc publish*(s: Switch, topic: string, data: seq[byte]): Future[void] =
   # pubslish to pubsub topic
   if s.pubSub.isNone:
-    raise newNoPubSubException()
+    var retFuture = newFuture[void]("Switch.publish")
```

Comment:
shouldn't there be a helper for creating a future with an exception? this  is repeated all over the place... ie `return newFutureExc("name", newException())`